Scala/binary-trees/BinaryTrees.scala

```diff
@@ -0,0 +1,75 @@
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
+
+final case class Tree(left: Tree, right: Tree) {
+  def checkSum: Int =
+    left match {
+      case null => 1
+      case tl   => 1 + tl.checkSum + right.checkSum
+    }
+}
+
+object Tree {
+  final val EmptyTree = Tree(null, null)
+
+  def sync(depth: Int): Tree = {
+    def innerBranch() = Tree.sync(depth - 1)
+    if (depth > 0) Tree(innerBranch(), innerBranch())
```

Comment:
I would inline the `def`.
```suggestion
    if (depth > 0) Tree(Tree.sync(depth - 1), Tree.sync(depth - 1))
```

Reply:
Do think that makes a difference?